### PR TITLE
reduce runtime of resolveSyntheticField

### PR DESCRIPTION
**Pull request by geoberle**
this implementation of the synthetic field resolver is not as clean as the previous one but runs a lot faster, which is crucial for the nodejs eventloop to not block.

### example 1
* simple subAttr field - query apps and their namespaces via synthetic relation
* old implementation: 245-300ms
* new implementation: 164-195ms

### example 2
* nested subAttr field - query saasfiles and their selfServiceRoles via the synthetic relation
* old implementation: 440-640ms
* new implementation: 320-375ms

this chagne is part of multiple approaches to unblock the nodejs event loop

Signed-off-by: Gerd Oberlechner <goberlec@redhat.com>

### CODE REVIEW

File: src/schema.ts [hunk 1]
```diff
@@ -75,6 +75,36 @@ const getGraphqlTypeForDatafileSchema = (app: express.Express, bundleSha: string
 // helpers
 const isNonEmptyArray = (obj: any) => obj.constructor === Array && obj.length > 0;
 
+const resolveSyntheticFieldInDatafile = (path: string, datafile: any,
```
Comment (janboll) [inline src/schema.ts:78]: Regarding name, as I understand, it's more of an exists function, perhaps rename to `resolveSyntheticFieldExists` ?

**Review by janboll — APPROVED** [review]

File: src/schema.ts (hunk 1 above)
Comment (geoberle) [reply src/schema.ts]: very good point! renamed it

File: src/schema.ts [hunk 2]
```diff
@@ -75,6 +75,36 @@ const getGraphqlTypeForDatafileSchema = (app: express.Express, bundleSha: string
 // helpers
 const isNonEmptyArray = (obj: any) => obj.constructor === Array && obj.length > 0;
 
+const pathRefExistsinDatafile = (path: string, datafile: any,
```
Comment (fishi0x01) [inline src/schema.ts:78]: Im not 100% sure why this saves time. My naive understanding is that we have less redundant map/filter calls here now? Is this something worth mentioning in a short comment, that this function was specifically designed to reduce the blocking time of the node event loop? Just so nobody gets the idea to add syntactic sugar back later :sweat_smile:

Comment (geoberle) [reply src/schema.ts]: :) added hint in [e686f3e](https://github.com/app-sre/qontract-server/pull/172/commits/e686f3e1c349b422aae1d741f49b1a080fe97c29)

**Review by fishi0x01 — APPROVED** [review]

**Pull request merged**